### PR TITLE
update company name from backend, remove hardcoded company name

### PR DESCRIPTION
**Pull request by creative-dev-lab**
**Issues during this PR**

During this PR, I noticed there can be a few improvements in optin widget.
So basically, it's mostly based on `UIView` with a few `UIStackView` combinations. but I found it can be fully designed by using `UIStackView` and I updated the UI based on `UIStackView`.
But during the updating UI, I noticed a few issues down below.

- [x] links are not clickable in the cart page of the demo
- [x] update optin widget ui by using stackview (not an issue, but improvement)
- [x] solve "By opting in you agree..." text height issue when expanded
- [x] "Wear now, resell later ..." text cutting off issue. So for this issue, It looks like there is enough space to show the full text, but "Wear now" is still cut off.
- [x] widget is not hidden when we remove an item and the remaining items are not eligible in the cart page of the demo

### Test plans

 **How it works**
We are currently getting company name, privacy url, find out more url, category mapping data etc. from `connector/graphql` with below query.
```
query GetShoppingData ($appId: String) {
  shop(appId: $appId) {
    name
    privacyUrl
    popupFindOutMoreUrl
    categoryMappings(mappedOrExcludedOnly: true) {
      from
      to
      excluded
    }
    minimumPriceThresholds {
      default
      department
      minimumPriceCents
    }
  }
}
```

if `privacyUrl` is `null`, we hide "By opting in..." text block.
if not, we use this `privacyUrl` for the privacy link in "By opting in..." text block.
Also we use `name` value for the company name in "By opting in..." text block.
And we use `popupFindOutMoreUrl` for the find out more URL in Learn More popup widget.

-- Note
`name` and `popupFindOutMoreUrl` are not nullable while `privacyUrl` is nullable.
And in order to test it successfully, `categoryMappings` should have data. but optional.

**Test Environment**
- Open our demo project through XCode.
- Inside `../AirRobeDemo/Application/AppDelegate -> AppDelegate class -> application()`, we have `initialize` function which has `appId` and `mode` as params.
- Update these `appId` and `mode` and test out the below cases.

**Test Cases**
We can test this out by simulating the connector api call with a specific `app_id`.
- [x] check whether "By opting in ..." text block is hidden if the privacy URL is `null`.
- [x] check whether it's updated with company name and privacy URL as per it comes from the connector based on `app_id`.
- [x] check whether it's updated with popup find more URL as per it comes from the connector based on `app_id`.

### CODE REVIEW

**Comment by piusj** [comment]
- check whether "By opting in ..." text block is hidden if the privacy URL is null.
**This doesn't work as expected. i set the sandbox app to nil and the text block still shows**

**Comment by piusj** [comment]
Its currently set to nil for that app so please test again

**Comment by piusj** [comment]
Unable to test this. Getting build errors on my local environment.

Can you try testing the flows again? The privacy_url of the app is null.

image

**Comment by egold** [comment]
@creative-dev-lab Random thought, but adding the Cart button to the NavigationBar on all screens would cut down on testing time.

**Comment by creative-dev-lab** [comment]
> @creative-dev-lab Random thought, but adding the Cart button to the NavigationBar on all screens would cut down on testing time.

That's a good point! 👍 
I will add the cart button to the navigation bar on all screens in the next PR for sure.

File: AirRobeDemo/AirRobeDemo/Application/AppDelegate.swift [hunk 1]
```diff
@@ -57,7 +57,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate, AirRobeEventDelegate {
         AirRobeWidget.initialize(
             config: AirRobeWidgetConfig(
                 appId: "515b6ee129da",
-                privacyPolicyURL: "https://www.example.com/privacy-policy",
```
Comment (egold) [inline AirRobeDemo/AirRobeDemo/Application/AppDelegate.swift:60]: Should probably updated `README.md` to match this.

**Comment by egold** [comment]
I'm still noticing layout constraint issues. Left unresolved, these can manifest as weird or hard to diagnose bugs in production. Can you please resolve these layout issues?

```
2022-07-12 18:11:56.136692-0400 AirRobeDemo[18542:5236325] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x60000155e940 'fittingSizeHTarget' UIStackView:0x7fe3e6a7c5f0.width == 0   (active)>",
    "<NSLayoutConstraint:0x60000155c5a0 'UISV-canvas-connection' UIStackView:0x7fe3e6a7c5f0.leading == UILabel:0x7fe3e6a7e560.leading   (active)>",
    "<NSLayoutConstraint:0x60000155cd20 'UISV-canvas-connection' H:[UIView:0x7fe3e6a7f330]-(0)-|   (active, names: '|':UIStackView:0x7fe3e6a7c5f0 )>",
    "<NSLayoutConstraint:0x60000155cd70 'UISV-spacing' H:[UILabel:0x7fe3e6a7e560]-(5)-[UILabel:0x7fe3e6a7e840]   (active)>",
    "<NSLayoutConstraint:0x60000155cdc0 'UISV-spacing' H:[UILabel:0x7fe3e6a7e840]-(5)-[UIActivityIndicatorView:0x7fe3e6a7ed40]   (active)>",
    "<NSLayoutConstraint:0x60000155ce10 'UISV-spacing' H:[UIActivityIndicatorView:0x7fe3e6a7ed40]-(5)-[UIView:0x7fe3e6a7f330]   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x60000155ce10 'UISV-spacing' H:[UIActivityIndicatorView:0x7fe3e6a7ed40]-(5)-[UIView:0x7fe3e6a7f330]   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
```

**Comment by creative-dev-lab** [comment]
> Unable to test this. Getting build errors on my local environment.
> 
> Can you try testing the flows again? The privacy_url of the app is null.
> 
> image

hey @piusj 
Sorry for the delayed response.
you should probably choose the device as iPhone, not an iMac. and this will solve the issue.

File: AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift [hunk 2]
```diff
@@ -20,6 +20,8 @@ final class CartPageViewController: UIViewController {
         super.viewDidLoad()
         initCart()
         initMultiOptIn()
+
+        emailTextField.delegate = self
```
[the remark this answers is not in the document]
Comment (creative-dev-lab) [reply AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift:24]: the library didn't do the email validation itself, and I believed it can be taken care of by the consumer.
But for sure, I can add email validation in another PR.
Thanks for pointing out this issue. 👍

**Comment by egold** [comment]
> 



> > Unable to test this. Getting build errors on my local environment.
> > Can you try testing the flows again? The privacy_url of the app is null.
> > image
> 
> hey @piusj Sorry for the delayed response. you should probably choose the device as iPhone, not an iMac. and this will solve the issue.

He's also compiling the SDK itself not the example app. @piusj was the intention to test the example app or test compilation/packaging of the SDK?

File: AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (egold) [reply AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift:24]: I am not sure it's the right thing to do, which is why I posed it as a question. For example, is the customer OK if we block the checkout flow due to an invalid email address for the AirRobe opt-in? Maybe check with Victor or whoever is the stakeholder for this aspect of this before acting on my question. :)

File: Sources/AirRobeWidget/Resources/AirRobeStrings.swift [hunk 3]
```diff
@@ -9,33 +9,33 @@ import Foundation
...
     // MARK: - Detailed Description Strings
-
     static let detailedDescription = "We’ve partnered with AirRobe to enable you to join the circular fashion movement. Re-sell or rent your purchases on AirRobe’s marketplace – all with one simple click. Together, we can do our part to keep fashion out of landfill. Learn more."
     static let learnMoreLinkText = "Learn more."
+    static let learnMoreLinkForPurpose = URL(string: "https://airrobe.com/")!
 
     // MARK: - Extra Info Strings
-
-    static let extraInfo = "By opting in you agree to THE ICONIC’s Privacy Policy and consent for us to share your details with AirRobe."
+    static let extraInfo = "By opting in you agree to COMPANY NAME’s Privacy Policy and consent for us to share your details with AirRobe."
```
Comment (egold) [inline Sources/AirRobeWidget/Resources/AirRobeStrings.swift:28]: Should probably use the variable `companyNameText` here instead of having a string literal on line 28 and on line 30. It's a bug waiting to happen in the future.

Safer: 

```swift
    static let companyNameText = "COMPANY NAME"
    static let extraInfo = "By opting in you agree to \(companyNameText)’s Privacy Policy and consent for us to share your details with AirRobe."
    static let extraLinkText = "Privacy Policy"

```

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 4]
```diff
@@ -13,8 +13,8 @@ import Combine
 final class AirRobeOptInView: UIView, NibLoadable {
     @IBOutlet weak var widgetStackView: UIStackView!
     @IBOutlet weak var mainContainerView: UIView!
-    @IBOutlet weak var mainContainerExpandButton: UIButton!
-    @IBOutlet weak var margin: UIView!
+    @IBOutlet weak var topContentContainer: UIStackView!
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:16]: Good to use a UIStackView; it's a good practice when usable space is unknown or variable (different screen sizes)

However, I wonder if perhaps the new view hierarchy is partly responsible for the layout constraint warnings? Might be worth investingating.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 5]
```diff
@@ -58,13 +58,13 @@ final class AirRobeOptInView: UIView, NibLoadable {
...
-        let maxWidth = subTitleContainer.bounds.width - potentialValueLabel.bounds.width - 10 - potentialValueLoading.bounds.width
-        if descriptionValueLabelMaxWidth != maxWidth && viewType == .optIn {
+        let maxWidth = subTitleContainer.bounds.width - potentialValueLabel.bounds.width - potentialValueLoading.bounds.width
+        if descriptionValueLabelMaxWidth != maxWidth {
             descriptionValueLabelMaxWidth = maxWidth
             guard let value = descriptionLabel.text, value.isEmpty || value == AirRobeStrings.description else {
                 return
             }
-            if (AirRobeStrings.description).width(withFont: descriptionLabel.font).width > descriptionValueLabelMaxWidth {
+            if ((AirRobeStrings.description).width(withFont: descriptionLabel.font).width + 20) > descriptionValueLabelMaxWidth {
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:67]: Please use a constant here instead of a magic number.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 6]
```diff
@@ -58,13 +58,13 @@ final class AirRobeOptInView: UIView, NibLoadable {
     }
 
     override func layoutSubviews() {
-        let maxWidth = subTitleContainer.bounds.width - potentialValueLabel.bounds.width - 10 - potentialValueLoading.bounds.width
-        if descriptionValueLabelMaxWidth != maxWidth && viewType == .optIn {
+        let maxWidth = subTitleContainer.bounds.width - potentialValueLabel.bounds.width - potentialValueLoading.bounds.width
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:61]: Perhaps a helper method here would help explain what this calculation is for?

For example:

```swift
let maxWidth = usableHorizontalSpace()
```

and having the calculation be hidden behind the helper method helps make the intention of the calculation more explicit.

Comment (egold) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:61]: It also allows for testing that function with a unit test at some point as well

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 7]
```diff
@@ -76,37 +76,42 @@ final class AirRobeOptInView: UIView, NibLoadable {
...
-            #if DEBUG
-            print("Privacy policy url is not valid.")
-            #endif
-            return
-        }
         detailedDescriptionLabel.setLinkText(
             orgText: AirRobeStrings.detailedDescription,
             linkText: AirRobeStrings.learnMoreLinkText,
-            link: privacyLink,
+            link: AirRobeStrings.learnMoreLinkForPurpose,
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:96]: Good improvement here

File: AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (creative-dev-lab) [reply AirRobeDemo/AirRobeDemo/Controllers/CartPageViewController.swift:24]: Cc @michaeldawson @piusj

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 8]
```diff
@@ -76,37 +76,42 @@ final class AirRobeOptInView: UIView, NibLoadable {
...
+                linkText: AirRobeStrings.extraLinkText,
+                link: privacyLink,
+                tapHandler: onTapExtraInfoLink)
+        } else {
+            extraInfoContainer.isHidden = true
+            #if DEBUG
+            print("Privacy policy url is not valid.")
+            #endif
+        }
+
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:114]: I find this block to be somewhat hard to follow because it is mixing multiple concerns -- if there is a privacyLink, then show the extraInfoContainer and set its link text based on the extraInfo / extraLinkText / privacyLink, along with this tapHandler. Otherwise, hide the extra info container.

Seeing multiple concerns in one block of code makes me wonder if privacyLink has too much responsibility? For example, what if there is extraInfo, extraLinkText, but there is no privacyLink? Does the lack of a privacyLink mean the extraInfo / extraLinkText should be ignored?

I wonder if this could be broken out a bit into a few bits of logic each with a single purpose?

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift [hunk 9]
```diff
@@ -273,7 +283,7 @@ private extension AirRobeOptInView {
                 DispatchQueue.main.async {
                     self.potentialValueLoading.stopAnimating()
                     self.potentialValueLabel.text = AirRobeStrings.potentialValue + "$" + price
-                    if ((AirRobeStrings.description).width(withFont: self.descriptionLabel.font).width + 15) < self.descriptionValueLabelMaxWidth {
+                    if ((AirRobeStrings.description).width(withFont: self.descriptionLabel.font).width + 20) > self.descriptionValueLabelMaxWidth {
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:286]: Can we replace the magic number `20` with a semantically named constant?

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInViewModel.swift [hunk 10]
```diff
@@ -58,11 +59,19 @@ final class AirRobeOptInViewModel {
...
+            return
+        }
+        let eligibility = AirRobeShoppingDataModelInstance.shared.categoryMapping.checkCategoryEligible(items: [category])
+        if isAllSet == .eligible {
+            if !(eligibility.eligible && !shoppingDataModel.isBelowPriceThreshold(department: department, price: priceCents)) {
+                isAllSet = .notEligible
+                AirRobeUtils.dispatchEvent(eventName: EventName.widgetNotRendered.rawValue, pageName: PageName.product.rawValue)
             }
-            let eligibility = AirRobeShoppingDataModelInstance.shared.categoryMapping.checkCategoryEligible(items: [category])
+        } else {
```
Comment (egold) [inline Sources/AirRobeWidget/Views/OptInView/AirRobeOptInViewModel.swift:74]: Does the category eligibility logic belong in a view model? It seems like a general data concern that isn't view specific?

Comment (egold) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInViewModel.swift:74]: I see that this is calling out to a couple of data model methods around price + category eligibility. The fact that this work is being done twice in this file makes me think it could be abstracted?

**Comment by piusj** [comment]
> > 
> 
> > > Unable to test this. Getting build errors on my local environment.
> > > Can you try testing the flows again? The privacy_url of the app is null.
> > 
> > 
> > hey @piusj Sorry for the delayed response. you should probably choose the device as iPhone, not an iMac. and this will solve the issue.
> 
> He's also compiling the SDK itself not the example app. @piusj was the intention to test the example app or test compilation/packaging of the SDK?

The intention was to test the example app. Thanks for this I'll give it another go

**Comment by piusj** [comment]
> > @creative-dev-lab Random thought, but adding the Cart button to the NavigationBar on all screens would cut down on testing time.
> 
> That's a good point! 👍 I will add the cart button to the navigation bar on all screens in the next PR for sure.

@creative-dev-lab Eli, in general we like to resolve issues in the PR they were reported. Could you make the changes in this PR rather than the next one? Just so we have the full context of the discussion here along with the changes. There are no urgent updates for the mobile SDKs, so I think we can keep this PR open and resolve Eric's recommendations here. It would be much easier for the team to follow this way.

File: Sources/AirRobeWidget/Resources/AirRobeStrings.swift (hunk 3 above)
Comment (creative-dev-lab) [reply Sources/AirRobeWidget/Resources/AirRobeStrings.swift:28]: Good point 👍 

> > > @creative-dev-lab Random thought, but adding the Cart button to the NavigationBar on all screens would cut down on testing time.
> > 
> > 
> > That's a good point! 👍 I will add the cart button to the navigation bar on all screens in the next PR for sure.
> 
> @creative-dev-lab Eli, in general we like to resolve issues in the PR they were reported. Could you make the changes in this PR rather than the next one? Just so we have the full context of the discussion here along with the changes. There are no urgent updates for the mobile SDKs, so I think we can keep this PR open and resolve Eric's recommendations here. It would be much easier for the team to follow this way.

Yep, no problem with that.

**Comment by creative-dev-lab** [comment]
> > > @creative-dev-lab Random thought, but adding the Cart button to the NavigationBar on all screens would cut down on testing time.
> > 
> > 
> > That's a good point! 👍 I will add the cart button to the navigation bar on all screens in the next PR for sure.
> 
> @creative-dev-lab Eli, in general we like to resolve issues in the PR they were reported. Could you make the changes in this PR rather than the next one? Just so we have the full context of the discussion here along with the changes. There are no urgent updates for the mobile SDKs, so I think we can keep this PR open and resolve Eric's recommendations here. It would be much easier for the team to follow this way.

Yep, no problem with that.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift (hunk 5 above)
Comment (creative-dev-lab) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:67]: Good point.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift (hunk 6 above)
Comment (creative-dev-lab) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:61]: This makes sense! 👍 
will create a helper method for this.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift (hunk 8 above)
Comment (creative-dev-lab) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInView.swift:114]: Yes, for now, we are hiding that extraInfo text block("By opting in...") if privacy url is `nil`.

File: Sources/AirRobeWidget/Views/OptInView/AirRobeOptInViewModel.swift (hunk 10 above)
Comment (creative-dev-lab) [reply Sources/AirRobeWidget/Views/OptInView/AirRobeOptInViewModel.swift:74]: Could you add more details?

**Review by maxckelly — APPROVED** [review]
